test/cli/provider_test.go

```diff
@@ -0,0 +1,148 @@
+			require.Equal(t, expectedProvider, res.Stdout.Trimmed())
+		}
+	}
+
+	t.Run("Basic Providing", func(t *testing.T) {
+		t.Parallel()
+
+		nodes := initNodes(t, 2, func(n *harness.Node) {
+			n.SetIPFSConfig("Experimental.StrategicProviding", false)
+		})
+		defer nodes.StopDaemons()
+
+		cid := nodes[0].IPFSAddStr(time.Now().String())
+		expectProviders(t, cid, nodes[0].PeerID().String(), nodes[1:]...)
+	})
+
+	t.Run("Basic Strategic Providing", func(t *testing.T) {
+		t.Parallel()
+
+		nodes := initNodes(t, 2, func(n *harness.Node) {
+			n.SetIPFSConfig("Experimental.StrategicProviding", true)
+		})
+		defer nodes.StopDaemons()
+
+		cid := nodes[0].IPFSAddStr(time.Now().String())
+		expectNoProviders(t, cid, nodes[1:]...)
+	})
+
+	t.Run("Reprovides with 'all' strategy", func(t *testing.T) {
+		t.Parallel()
+
+		nodes := initNodes(t, 2, func(n *harness.Node) {
+			n.SetIPFSConfig("Reprovider.Strategy", "all")
+		})
+		defer nodes.StopDaemons()
+
+		cid := nodes[0].IPFSAddStr(time.Now().String(), "--local")
+
+		expectNoProviders(t, cid, nodes[1:]...)
+
+		nodes[0].IPFS("bitswap", "reprovide")
+
+		expectProviders(t, cid, nodes[0].PeerID().String(), nodes[1:]...)
+	})
+
+	t.Run("Reprovides with 'pinned' strategy", func(t *testing.T) {
+		t.Parallel()
+
+		foo := testutils.RandomBytes(1000)
+		bar := testutils.RandomBytes(1000)
+
+		nodes := initNodes(t, 2, func(n *harness.Node) {
+			n.SetIPFSConfig("Reprovider.Strategy", "pinned")
+		})
+		defer nodes.StopDaemons()
+
+		cidFoo := nodes[0].IPFSAdd(bytes.NewReader(foo), "--offline", "--pin=false")
+		cidBar := nodes[0].IPFSAdd(bytes.NewReader(bar), "--offline", "--pin=false")
+		cidBarDir := nodes[0].IPFSAdd(bytes.NewReader(bar), "-Q", "--offline", "-w")
+
+		expectNoProviders(t, cidFoo, nodes[1:]...)
+		expectNoProviders(t, cidBar, nodes[1:]...)
+		expectNoProviders(t, cidBarDir, nodes[1:]...)
+
+		nodes[0].IPFS("bitswap", "reprovide")
+
+		expectNoProviders(t, cidFoo, nodes[1:]...)
+		expectProviders(t, cidBar, nodes[0].PeerID().String(), nodes[1:]...)
+		expectProviders(t, cidBarDir, nodes[0].PeerID().String(), nodes[1:]...)
+	})
+
+	t.Run("Reprovides with 'roots' strategy", func(t *testing.T) {
+		t.Parallel()
+
+		foo := testutils.RandomBytes(1000)
+		bar := testutils.RandomBytes(1000)
+		baz := testutils.RandomBytes(1000)
+
+		nodes := initNodes(t, 2, func(n *harness.Node) {
+			n.SetIPFSConfig("Reprovider.Strategy", "roots")
+		})
+		defer nodes.StopDaemons()
+
+		cidFoo := nodes[0].IPFSAdd(bytes.NewReader(foo), "--offline", "--pin=false")
+		cidBar := nodes[0].IPFSAdd(bytes.NewReader(bar), "--offline", "--pin=false")
+		cidBaz := nodes[0].IPFSAdd(bytes.NewReader(baz), "--offline")
+		cidBarDir := nodes[0].IPFSAdd(bytes.NewReader(bar), "-Q", "--offline", "-w")
+
+		expectNoProviders(t, cidFoo, nodes[1:]...)
+		expectNoProviders(t, cidBar, nodes[1:]...)
+		expectNoProviders(t, cidBarDir, nodes[1:]...)
+
+		nodes[0].IPFS("bitswap", "reprovide")
```

Comment:
:thought_balloon: fysa  moving  this to `ipfs routing reprovide` and deprecate `bitswap` one is tracked in unrelated: mind filling issue to

Reply:
https://github.com/ipfs/kubo/issues/10387